### ISSUE TITLE
Some problems with XuggleVideo

### ISSUE THREAD

**Issue by jonhare** [openimaj/openimaj#31]
OP-32: XuggleVideo has a few problems. To summarise:
- VideoDisplay on XuggleVideos do not loop correctly. 
- XuggleVideos used in isolation (i.e. for(MBFImage img : video) never end as hasNextFrame() lies (i.e. returns true when there are no further frames). This breaks on 2 video files we've played with
- rtsp:// urls don't seem to work

**Comment by jonhare** [comment]
Original reporter: @sinjax

**Comment by jonhare** [comment]
sinjax:  I've attached a file containing a test that definitely breaks with XuggleVideo how it is

**Comment by jonhare** [comment]
davedupplaw:  So, there was a problem with the test - the FileUtils.copyStreamToFile() method assumes a text file and is corrupting the video. I'll commit a method to the FileUtils which will work with binary streams.  

The problem of looping (seeking to 0 at the end) seems to be an issue with Xuggler 5.4. The seekToKeyFrame method seems really flakey and doesn't actually seem to do its job properly.  I'm going to try adding a seek method that will check the timestamp after its seeked and use multiple seek types to attempt to get as close to the real timestamp as possible, but it might be slower.

I've found the problem with the hasNextFrame() thing that was lying. I'll commit that later too.

Haven't looked at rtsp streams yet. Do you have a stream that you tested it on that wasn't working?

**Comment by jonhare** [comment]
jonhare:  Not RTSP, but mjpeg:

final XuggleVideo xv = new XuggleVideo("http://137.229.91.244/mjpg/video.mjpg");
VideoDisplay.createVideoDisplay(xv);

Interestingly the following does seem to be getting frames:
final IMediaReader reader = ToolFactory.makeReader("http://137.229.91.244/mjpg/video.mjpg");

reader.addListener(new MediaListenerAdapter() {
	@Override
	public void onVideoPicture(final IVideoPictureEvent event) {
		System.out.println("Got frame" + event.getPicture().getSize());
	}
});

while (reader.readPacket() == null)
	;

**Comment by jonhare** [comment]
jonhare:  final XuggleVideo xv = new XuggleVideo("rtsp://ipeyeanywhere.marmitek.com/mpeg4");

results in:

java.io.FileNotFoundException: rtsp:/ipeyeanywhere.marmitek.com/mpeg4 (No such file or directory)
	at java.io.RandomAccessFile.open(Native Method)
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:216)
	at java.io.RandomAccessFile.<init>(RandomAccessFile.java:101)
	at org.openimaj.video.xuggle.XuggleVideo.create(XuggleVideo.java:444)
	at org.openimaj.video.xuggle.XuggleVideo.<init>(XuggleVideo.java:302)
	at org.openimaj.video.xuggle.XuggleVideo.<init>(XuggleVideo.java:259)
	at org.openimaj.Test.main(Test.java:15)
Maybe not a URL? : 'rtsp://ipeyeanywhere.marmitek.com/mpeg4'
Nope, not a file either. *shrug*
Video ended

So presumably there is a problem with the URL scheme?

**Comment by jonhare** [comment]
davedupplaw:  The updated test that Sina attached.

**Comment by jonhare** [comment]
davedupplaw:  I've attached the updated test of Sina's that I've been running.

I've updated XuggleVideo with a seekToBeginning() method (the method reset() calls it) that tries all Xuggler's various methods to get a container to reset itself. It seems to work on the files I've tried.

I've updated XuggleVideo with a new method for creating the container that tries to use Xuggler's built in methods first rather than second. It then falls back to using a Java InputStream (for resources that are in Jars). This seems to work for the URL test Jon has given.

I've tested the RTSP url with rtsp://184.72.239.149/vod/mp4:BigBuckBunny_115k.mov and that works fine too. The given RTSP URL wasn't working for me.
